src/System.Management.Automation/engine/parser/tokenizer.cs

```diff
@@ -3608,16 +3720,34 @@ private string ScanNumberHelper(char firstChar, out bool hex, out bool real, out
             else
             {
                 c = PeekChar();
-                if (firstChar == '0' && (c == 'x' || c == 'X'))
+                bool isHexOrBinary = firstChar == '0' && (c == 'x' || c == 'X' || c == 'b' || c == 'B');
+                if (isHexOrBinary)
                 {
                     SkipChar();
-                    ScanHexDigits(sb);
-                    if (sb.Length == 0)
+                    switch (c)
                     {
-                        notNumber = true;
-                    }
+                        case 'x':
```

Comment:
Above in line 3723 we have checked this - we should remove the duplication.

Reply:
🤔 Okay, removing this duplication is a mite tricky... Think I got something workable though!

Reply:
At first glance, it was a simple change. If not, then maybe we shouldn't complicate the code.

Reply:
Yep, seems like that ought to work. Not the biggest fan of jumping around like that, but it seems to work and avoids both the `bool` allocation as well as the unnecessary double-check on any given traversal.

Reply:
It was mainly a bit difficult because of how the code path was originally structured, and the if/else condition. I think the current code I have here should do the job without the double-checks. 🙂